File: boosting/src/learning/classification/ClassificationModel.java
package learning.classification;

import data.ClassifiedData;
import data.Data;
import data.RegressionData;
import data.RegressionDataDecorator;
import learning.model.ModelWithTeacher;

import java.util.ArrayList;
import java.util.List;
import java.util.function.Function;

public class ClassificationModel implements Classifier {
    private final ModelWithTeacher model;
    private final Function<double[], Integer> outputToClass;

    public ClassificationModel(ModelWithTeacher model, Function<double[], Integer> outputToClass) {
        this.model = model;
        this.outputToClass = outputToClass;
    }

    @Override
    public int classify(Data data) {
        return outputToClass.apply(output(data));
    }

    public void trainClassified(List<ClassifiedData> data) {
        List<RegressionData> regressionData = new ArrayList<>();
        for (ClassifiedData classifiedData : data) {
            regressionData.add(new RegressionDataDecorator(classifiedData));
        }
        model.trainAll(regressionData);
    }

    @Override
    public void train(RegressionData data) {
        model.train(data);
    }

    @Override
    public void trainAll(List<? extends RegressionData> data) {
        throw new RuntimeException("Classification model should be trained on class-labeled data");
    }

    @Override
    public double[] output(Data data) {
        return model.output(data);
    }
}

File: boosting/src/learning/classification/ClassificationModel.kt
package learning.classification

import data.ClassifiedData
import data.Data
import data.RegressionData
import data.RegressionDataDecorator
import learning.model.ModelWithTeacher
import java.util.*

class ClassificationModel(private val model: ModelWithTeacher, private val outputToClass: (DoubleArray) -> Int) : Classifier {

    override fun classify(data: Data): Int {
        return outputToClass(output(data))
    }

    fun trainClassified(data: List<ClassifiedData>) {
        val regressionData = ArrayList<RegressionData>()
        for (classifiedData in data) {
            regressionData.add(RegressionDataDecorator(classifiedData))
        }
        model.trainAll(regressionData)
    }

    override fun train(data: RegressionData) {
        model.train(data)
    }

    override fun trainAll(data: List<RegressionData>) {
        throw RuntimeException("Classification model should be trained on class-labeled data")
    }

    override fun output(data: Data): DoubleArray {
        return model.output(data)
    }
}
